### PR TITLE
FEAT-#626: Add `to_parquet` for Ray execution engine

### LINKED ISSUE #626: to_parquet is not supported
### System information

```
$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.2 LTS"
$ conda --version
conda 4.6.14
$ python --version
Python 3.7.3
$ pip --version
pip 19.1 from /home/dlweber/miniconda3/envs/gis-dataprocessing/lib/python3.7/site-packages/pip (python 3.7)

$ pip freeze | grep modin
modin==0.5.0
$ pip freeze | grep pandas
pandas==0.24.2
$ pip freeze | grep numpy
numpy==1.16.3
```

miniconda3 was used to install most of the sci-py stack, with a pip clause to add modin, e.g.
```
# environment.yaml
channels:
  - conda-forge
  - defaults

dependencies:
  - python>=3.7
  - affine
  - configobj
  - dask
  - numpy
  - pandas
  - pyarrow
  - rasterio
  - s3fs
  - scikit-learn
  - scipy
  - shapely
  - xarray
  - pip
  - pip:
    - modin
```

### Describe the problem

https://modin.readthedocs.io/en/latest/pandas_supported.html says `to_parquet` is supported, but maybe not:

```
import numpy as np
import modin.pandas as pd
size = (1, 10 * 10)
column_ij = ["%04d_%04d" % (i, j) for i in range(10) for j in range(10)]
data = np.random.randint(0, 10000, size=size, dtype="uint16")
df = pd.DataFrame(data, columns=column_ij)
df.to_parquet('/tmp/tmp.parquet')
```

```
UserWarning: `DataFrame.to_parquet` defaulting to pandas implementation.
```

More details:
```
2019-05-21 16:03:46,207	WARNING worker.py:1337 -- WARNING: Not updating worker name since `setproctitle` is not installed. Install this with `pip install setproctitle` (or ray[debug]) to enable monitoring of worker processes.
2019-05-21 16:03:46,207	INFO node.py:469 -- Process STDOUT and STDERR is being redirected to /tmp/ray/session_2019-05-21_16-03-46_18437/logs.
2019-05-21 16:03:46,310	INFO services.py:407 -- Waiting for redis server at 127.0.0.1:55558 to respond...
2019-05-21 16:03:46,418	INFO services.py:407 -- Waiting for redis server at 127.0.0.1:41726 to respond...
2019-05-21 16:03:46,420	INFO services.py:804 -- Starting Redis shard with 2.1 GB max memory.
2019-05-21 16:03:46,426	INFO node.py:483 -- Process STDOUT and STDERR is being redirected to /tmp/ray/session_2019-05-21_16-03-46_18437/logs.
2019-05-21 16:03:46,427	WARNING services.py:1304 -- WARNING: The object store is using /tmp instead of /dev/shm because /dev/shm has only 5238738944 bytes available. This may slow down performance! You may be able to free up space by deleting files in /dev/shm or terminating any running plasma_store_server processes. If you are inside a Docker container, you may need to pass an argument with the flag '--shm-size' to 'docker run'.
2019-05-21 16:03:46,427	INFO services.py:1427 -- Starting the Plasma object store with 6.0 GB memory using /tmp.
UserWarning: Distributing <class 'list'> object. This may take some time.
UserWarning: `DataFrame.to_parquet` defaulting to pandas implementation.
To request implementation, send an email to feature_requests@modin.org.
```

Maybe modin could be added to conda-forge so that conda can help with resolving version dependencies?

### PR DESCRIPTION
**Pull request by naren-ponder**
Signed-off-by: Naren Krishna <naren@ponder.io>

## What do these changes do?

**DRAFT PR IN PROGRESS**

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #626  
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/execution/ray/generic/io/io.py [hunk 1]
```diff
@@ -196,3 +197,77 @@ def func(df, **kw):
         for rows in result:
             for partition in rows:
                 wait([partition.oid])
+
+    @staticmethod
+    def _to_parquet_check_support(kwargs):
+        """
+        Check if parallel version of ``to_parquet`` could be used.
```
Comment (YarShev) [inline modin/core/execution/ray/generic/io/io.py:204]:
```suggestion
        Check if parallel version of `to_parquet` could be used.
```
We use one back stick for methods and two ones for classes.

File: modin/core/execution/ray/generic/io/io.py [hunk 2]
```diff
@@ -196,3 +197,77 @@ def func(df, **kw):
...
+
+        Parameters
+        ----------
+        kwargs : dict
+            Keyword arguments passed to ``.to_parquet()``.
```
Comment (YarShev) [inline modin/core/execution/ray/generic/io/io.py:209]:
```suggestion
            Keyword arguments passed to `.to_parquet()`.
```

File: modin/core/execution/ray/generic/io/io.py [hunk 3]
```diff
@@ -196,3 +197,77 @@ def func(df, **kw):
...
+
+        Returns
+        -------
+        bool
+            Whether parallel version of ``to_parquet`` is applicable.
```
Comment (YarShev) [inline modin/core/execution/ray/generic/io/io.py:214]:
```suggestion
            Whether parallel version of `to_parquet` is applicable.
```

File: modin/core/execution/ray/generic/io/io.py [hunk 4]
```diff
@@ -196,3 +197,77 @@ def func(df, **kw):
...
+        if not isinstance(path, str):
+            return False
+        if any((path.endswith(ext) for ext in [".gz", ".bz2", ".zip", ".xz"])):
+            return False
+        if compression is None or not compression == "snappy":
+            return False
+        return True
+
+    @classmethod
+    def to_parquet(cls, qc, **kwargs):
```
Comment (YarShev) [inline modin/core/execution/ray/generic/io/io.py:227]: We should probably add a parallel version of `to_parquet` for Dask engine too since we have parallel `read_parquet` both for Ray and Dask.

File: modin/core/execution/ray/generic/io/io.py [hunk 5]
```diff
@@ -196,3 +197,77 @@ def func(df, **kw):
...
+        """
+        Write a ``DataFrame`` to the binary parquet format.
+
+        Parameters
+        ----------
+        qc : BaseQueryCompiler
+            The query compiler of the Modin dataframe that we want to run ``to_parquet`` on.
```
Comment (YarShev) [inline modin/core/execution/ray/generic/io/io.py:234]:
```suggestion
            The query compiler of the Modin dataframe that we want to run `to_parquet` on.
```

File: modin/core/execution/ray/generic/io/io.py [hunk 6]
```diff
@@ -196,3 +197,77 @@ def func(df, **kw):
...
+        **kwargs : dict
+            Parameters for ``pandas.to_parquet(**kwargs)``.
```
Comment (YarShev) [inline modin/core/execution/ray/generic/io/io.py:236]:
```suggestion
            Parameters for `pandas.to_parquet(**kwargs)`.
```

File: modin/core/execution/ray/generic/io/io.py [hunk 7]
```diff
@@ -196,3 +197,77 @@ def func(df, **kw):
...
+        def func(df, **kw):
+            """
+            Dump a chunk of rows as parquet, then save them to target maintaining order.
+
+            Parameters
+            ----------
+            df : pandas.DataFrame
+                A chunk of rows to write to a parquet file.
+            **kw : dict
+                Arguments to pass to ``pandas.to_parquet(**kw)`` plus an extra argument
```
Comment (YarShev) [inline modin/core/execution/ray/generic/io/io.py:250]:
```suggestion
                Arguments to pass to `pandas.to_parquet(**kwargs)` plus an extra argument
```

File: modin/core/execution/ray/generic/io/io.py [hunk 8]
```diff
@@ -196,3 +197,77 @@ def func(df, **kw):
...
+        ----------
+        qc : BaseQueryCompiler
+            The query compiler of the Modin dataframe that we want to run ``to_parquet`` on.
+        **kwargs : dict
+            Parameters for ``pandas.to_parquet(**kwargs)``.
+        """
+        if not cls._to_parquet_check_support(kwargs):
+            return BaseIO.to_parquet(qc, **kwargs)
+
+        def func(df, **kw):
```
Comment (YarShev) [inline modin/core/execution/ray/generic/io/io.py:241]: Where are `**kw`` used?

File: modin/pandas/dataframe.py [hunk 9]
```diff
@@ -2195,23 +2195,26 @@ def to_parquet(
         compression="snappy",
         index=None,
         partition_cols=None,
-        storage_options: StorageOptions = None,
```
Comment (YarShev) [inline modin/pandas/dataframe.py:2198]: Why is this removed?

File: modin/pandas/test/test_io.py [hunk 10]
```diff
@@ -95,6 +96,33 @@ def setup_clipboard(row_size=NROWS):
     df.to_clipboard()
 
 
+def parquet_eval_to_file(modin_obj, pandas_obj, fn, extension, **fn_kwargs):
+    """Helper function to test `to_<extension>` methods.
```
Comment (YarShev) [inline modin/pandas/test/test_io.py:100]: The docstring doesn't correspond to numpydoc.

File: modin/pandas/test/test_io.py [hunk 11]
```diff
@@ -1259,7 +1287,7 @@ def test_to_parquet(self):
         reason="The reason of tests fail in `cloud` mode is unknown for now - issue #3264",
     )
     def test_read_parquet_2462(self):
-        test_df = pd.DataFrame(
+        test_df = pandas.DataFrame(
```
Comment (YarShev) [inline modin/pandas/test/test_io.py:1290]: Why is this replaced to pandas usage?

File: modin/core/execution/ray/generic/io/io.py [hunk 12]
```diff
@@ -196,3 +197,77 @@ def func(df, **kw):
...
+            partition_idx = kw["partition_idx"]
+            if not os.path.exists(output_path):
+                os.makedirs(output_path)
+            kwargs[
+                "path"
+            ] = f"{output_path}/part-{partition_idx:04d}.{compression}.parquet"
+            df.to_parquet(**kwargs)
+            return pandas.DataFrame()
+
+        # signaling that the partition with id==0 can be written to the file
```
Comment (YarShev) [inline modin/core/execution/ray/generic/io/io.py:264]: I guess you copied the code of recent implementation of `read_csv`. Please, make sure that parallel `to_parquet` is correct.

File: modin/core/execution/ray/generic/io/io.py (hunk 8 above)
Comment (naren-ponder) [reply modin/core/execution/ray/generic/io/io.py:241]: It is used to get the partition_idx for the output file name.

File: modin/pandas/dataframe.py (hunk 9 above)
Comment (naren-ponder) [reply modin/pandas/dataframe.py:2198]: I see that the default is None is the Pandas docs here (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_parquet.html?highlight=to_parquet#pandas.DataFrame.to_parquet), did not understand the need for adding the `StorageOptions`

File: modin/pandas/test/test_io.py (hunk 11 above)
Comment (naren-ponder) [reply modin/pandas/test/test_io.py:1290]: We are testing `read_parquet` here, so it follows as a test that we test the read on a Pandas Dataframe because the `read_parquet` method is Modin's.

File: modin/core/execution/ray/generic/io/io.py (hunk 12 above)
Comment (naren-ponder) [reply modin/core/execution/ray/generic/io/io.py:264]: Yes, `to_parquet` is writing to files in parallel and I have tested both manually and the ray tests on TeamCity pass.

File: modin/core/execution/ray/generic/io/io.py (hunk 4 above)
Comment (devin-petersohn) [reply modin/core/execution/ray/generic/io/io.py:227]: @YarShev it can happen in a different PR. I think it is probably better. to separate them

File: modin/pandas/test/test_io.py (hunk 11 above)
Comment (devin-petersohn) [reply modin/pandas/test/test_io.py:1290]: To expand on what @naren-ponder said, the new Modin `to_parquet` does not reproduce the bug that we fixed in #2462. It was incorrect written at first, but not caught because Modin was using pandas `to_parquet`

File: modin/core/execution/ray/generic/io/io.py (hunk 12 above)
Comment (devin-petersohn) [reply modin/core/execution/ray/generic/io/io.py:264]: @YarShev it is well tested at this point.

File: modin/core/execution/ray/generic/io/io.py (hunk 8 above)
Comment (devin-petersohn) [reply modin/core/execution/ray/generic/io/io.py:241]: It is documented below, I think it is ok. Do you have any issues @YarShev?

File: modin/pandas/dataframe.py (hunk 9 above)
Comment (devin-petersohn) [reply modin/pandas/dataframe.py:2198]: @naren-ponder We can probably add it back. The actual implementation of pandas uses it as a type hint: https://github.com/pandas-dev/pandas/blob/v1.3.4/pandas/core/frame.py#L2586

I think it probably should be `Optional[StorageOptions]` in pandas, but they choose to make it optional all the time which is where I think the confusion is coming from: https://github.com/pandas-dev/pandas/blob/v1.3.4/pandas/_typing.py#L177

File: modin/core/execution/ray/generic/io/io.py (hunk 4 above)
Comment (YarShev) [reply modin/core/execution/ray/generic/io/io.py:227]: I don't mind, but the PR title needs to be corrected.

File: modin/core/execution/ray/generic/io/io.py (hunk 8 above)
Comment (YarShev) [reply modin/core/execution/ray/generic/io/io.py:241]: Okay, I see

File: modin/core/execution/ray/generic/io/io.py (hunk 12 above)
Comment (YarShev) [reply modin/core/execution/ray/generic/io/io.py:264]: What is the comment above for if we don't perform any synchronization. Actually, why don't we need to perform the synchronization?

File: modin/pandas/dataframe.py (hunk 9 above)
Comment (YarShev) [reply modin/pandas/dataframe.py:2198]: Suggest adhering to pandas.

File: modin/pandas/test/test_io.py [hunk 13]
```diff
@@ -95,6 +96,33 @@ def setup_clipboard(row_size=NROWS):
     df.to_clipboard()
 
 
+def parquet_eval_to_file(modin_obj, pandas_obj, fn, extension, **fn_kwargs):
+    """Helper function to test `to_parquet` method.
```
Comment (YarShev) [inline modin/pandas/test/test_io.py:100]: The docstring is not numpydoc-like style yet. Please, fix it. You can look at how you document `to_parquet` as an example. https://github.com/modin-project/modin/pull/3573/files#diff-d6ab859ea39dde23e5d8b784b25150c9cf1b52b7f7d9233b99fd355145127130R228

File: modin/pandas/test/test_io.py (hunk 11 above)
Comment (YarShev) [reply modin/pandas/test/test_io.py:1290]: ok, makes sense to me

File: modin/core/execution/ray/generic/io/io.py (hunk 12 above)
Comment (naren-ponder) [reply modin/core/execution/ray/generic/io/io.py:264]: Ah that comment is not necessary, removed it. Parquet is a serialized way of storing a Pandas DataFrame (i.e. you are writing the exact representation from the dataframe to disk so the types of columns and indices are the same). So a synchronization process for different partitions is not necessary.

Comment (YarShev) [reply modin/core/execution/ray/generic/io/io.py]: Okay, I see

File: modin/pandas/test/test_io.py (hunk 13 above)
Comment (YarShev) [reply modin/pandas/test/test_io.py:101]: We have `eval_to_file` to check `to_*` functions. Why do we need this specific function for `to_parquet`?

File: modin/pandas/dataframe.py [hunk 14]
```diff
@@ -2197,21 +2197,24 @@ def to_parquet(
...
+
+        config = {
+            "path": path,
+            "engine": engine,
+            "compression": compression,
+            "index": index,
+            "partition_cols": partition_cols,
+            "storage_options": storage_options,
+        }
+        new_query_compiler = self._query_compiler
```
Comment (YarShev) [inline modin/pandas/dataframe.py:2210]: What is this assignment for?

File: modin/pandas/test/test_io.py (hunk 13 above)
Comment (naren-ponder) [reply modin/pandas/test/test_io.py:101]: Because modin to_parquet implementation creates multiple `.parquet` files (one for each partition) that needs to be checked against pandas which creates a single `.parquet` file. The other `to_*` modin implementations create a single file that can be checked against pandas, but for `to_parquet` we have to have specific code to test reading and writing entire directories and checking them against a single. Additionally, tearing these directories/files down at the end of the test requires specific code not handled by `eval_to_file`.

File: modin/pandas/dataframe.py (hunk 14 above)
Comment (naren-ponder) [reply modin/pandas/dataframe.py:2210]: This is similar to to_csv where we get the query compiler from the class itself then pass it to the Factory Dispatcher. In theory, `self._query_compiler` could be passed directly to the factory dispatcher without this assignment, but I think this is a bit cleaner and is similar to other portions of the codebase.

Comment (YarShev) [reply modin/pandas/dataframe.py:2210]: Okay, I see

File: modin/pandas/test/test_io.py (hunk 13 above)
Comment (YarShev) [reply modin/pandas/test/test_io.py:101]: makes sense

File: modin/pandas/test/test_io.py [hunk 15]
```diff
@@ -96,6 +97,38 @@ def setup_clipboard(row_size=NROWS):
...
+    ----------
+        modin_obj: pd.DataFrame
+            Modin DataFrame or Series to test `to_parquet` method.
+        pandas_obj: pandas.DataFrame
+            Pandas DataFrame or Series to test `to_parquet` method.
+        fn: str
+            Name of the method, that should be tested.
+        extension: str
+            Extension of the test file.
+    """
```
Comment (YarShev) [inline modin/pandas/test/test_io.py:113]:
```suggestion
    """
    Helper function to test `to_parquet` method.

    Parameters
    ----------
    modin_obj : pd.DataFrame
        Modin DataFrame or Series to test `to_parquet` method.
    pandas_obj: pandas.DataFrame
        Pandas DataFrame or Series to test `to_parquet` method.
    fn : str
        Name of the method, that should be tested.
    extension : str
        Extension of the test file.
    """
```
Since we do not check test files for docstrings CI passes but the docstring doesn't correspond yet to our guide. We should probably fix all docstrings in tests at some point.

**Review by YarShev — COMMENTED** [review]
Left a comment. All the rest looks good.

**Review by YarShev — APPROVED** [review]
@naren-ponder , LGTM, thanks!

**Review by devin-petersohn — APPROVED** [review]
LGTM, thanks for getting this in @naren-ponder!

File: modin/pandas/test/test_io.py (hunk 13 above)
Comment (anmyachev) [reply modin/pandas/test/test_io.py]: @devin-petersohn @YarShev if the behavior of the Modin function differs from the behavior of the pandas function, we should add it to the experimental mode, just as it was with `read_pickle_distributed` and `to_pickle_distributed` functions. Why is it done differ in this case?

Comment (devin-petersohn) [reply modin/pandas/test/test_io.py]: This is still consistent with parquet format, where `read_pickle_distributed` is not consistent with the format. We can still use pandas to read the output of this `parquet` writer, but it wouldn't work to read the directory of pickle files created by `read_pickle_distributed`.

Comment (anmyachev) [reply modin/pandas/test/test_io.py]: got it, thank you
